### DISCUSSION TITLE
prisma data-proxy

### DISCUSSION THREAD

**Discussion by marco-mastella** [prisma/prisma#18856] [Q&A]
### Bug description

when i try to deploy my new google cloud functions I receive the below error
I tried to run `npx prisma generate --data-proxy` many times

on my code I have
```
import { PrismaClient } from '@prisma/client/edge';
export const dbClient = new PrismaClient()
```

```
i  functions: updating Node.js 18 function api(europe-west3)...
i  functions: updating Node.js 18 function registerUser(europe-west3)...
i  functions: updating Node.js 18 function helloWorld(europe-west3)...
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Provided module can't be loaded.
Is there a syntax error in your code?
Detailed stack trace: Error: @prisma/client/edge did not initialize yet. Please run "prisma 
generate --data-proxy" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/workspace/node_modules/.prisma/client/edge.js:3:11)
    at Object.<anonymous> (/workspace/lib/utils/utils.js:16:20)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (/workspace/lib/users/controller.js:9:17)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
Could not load the function, shutting down.
. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.
Function failed on loading user code. This is likely due to a bug in the user code. Error message: Provided module can't be loaded.
Is there a syntax error in your code?
Detailed stack trace: Error: @prisma/client/edge did not initialize yet. Please run "prisma 
generate --data-proxy" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/workspace/node_modules/.prisma/client/edge.js:3:11)
    at Object.<anonymous> (/workspace/lib/utils/utils.js:16:20)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (/workspace/lib/users/controller.js:9:17)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
Could not load the function, shutting down.
. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation.

Functions deploy had errors with the following functions:
        api(europe-west3)
        helloWorld(europe-west3)
        registerUser(europe-west3)
i  functions: cleaning up build files...
Error: There was an error deploying functions:
- Error Failed to update function registerUser in region europe-west3
- Error Failed to update function helloWorld in region europe-west3
- Error Failed to update function api in region europe-west3
```

if I use 
```
import { PrismaClient } from '@prisma/client';
export const dbClient = new PrismaClient()
```
it build and deploy the functions but it raises the error on runtime 

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: Windows
- Database: Cloud SQL PostgreSQL 
- Node.js version: 18.7.0
- prisma version 4.13.0

### Prisma Version
- prisma version 4.13.0
```

```

**Comment by janpio** [comment]
How are you deploying your app to Google Cloud Functions?
What error are you getting in runtime when using the normal Client? Can't be the same as that would mean it contains an edge Client error string...

**Reply by marco-mastella** [reply]
Firstly, my question is: which is the correct configuration for a google cloud function infrastructure?
Is it with "/edge" or without?
If I check the connection on Prisma data proxy, I can see the data, and the connections string was copy&pasted on the code, so I would expect it is correct.

**Reply by janpio** [reply]
Both can work, see e.g. https://github.com/prisma/ecosystem-tests/tree/dev/platforms-serverless/gcp-functions for a setup that runs normal Prisma Client with a remote database. But for the Edge Client, the deployed Client has to be generated with `--data-proxy`.

The error message indicates that the Client you try to deploy was not. (Hence my question how you are deploying...)

**Reply by nechmads** [reply]
I get the same thing.
When running using data proxy the above error always comes up. When moving to use a connection string directly to the DB everything is working fine.
Data proxy works when running locally, but fail when running in Cloud Funcsions environment.

**Reply by janpio** [reply]
How are you deploying?

**Reply by nechmads** [reply]
hi @janpio. Deploying using firebase cli.
Basically:
- Using edge and data proxy works when running everything locally using the Firebase functions emulator.
- When deploying to GCP, I get thee errors above.
- When reverting to use a direct database connection string and redeploying to GCP everything works fine

**Reply by janpio** [reply]
That means the Client did definitely not get generated with `--data-proxy` there. Are you sure you ran that command in that environment? Does something else maybe run `npm install` _afterwards_ which in some way could overwrite your generate Client? (If that might be the case, try configuring the following env var in that environment: `PRISMA_GENERATE_DATAPROXY=true`)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
